fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,523 @@
+                ret = true;
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * generate a new BinaryPredicate.
+     * @return new BinaryPredicate.
+     */
+    private Expr genNewBinaryPredicate(Expr oldExpr, Expr newSlot) {
+        if (oldExpr instanceof BinaryPredicate) {
+            BinaryPredicate oldBP = (BinaryPredicate) oldExpr;
+            BinaryPredicate newBP = new BinaryPredicate(oldBP.getOp(), newSlot, oldBP.getChild(1));
+            return newBP;
+        }
+        return oldExpr;
+    }
+
+    /**
+     * add the new BinaryPredicate to slotToLiteralDeDuplication and newExprWithState and simultaneous deduplication
+     */
+    private void addNewBinaryPredicate(Expr expr,
+                                       Set<Pair<Expr, Expr>> slotToLiteralDeDuplication,
+                                       List<Pair<Expr, Boolean> > newExprWithState,
+                                       boolean needAddnewExprWithState,
+                                       Analyzer analyzer,
+                                       ExprRewriter.ClauseType clauseType) {
+        if (expr instanceof BinaryPredicate) {
+            BinaryPredicate newBP = (BinaryPredicate) expr;
+            Pair<Expr, Expr> pair = new Pair<>(newBP.getChild(0), newBP.getChild(1));
+            if (!slotToLiteralDeDuplication.contains(pair)) {
+                slotToLiteralDeDuplication.add(pair);
+                Pair<Expr, Boolean> newBPWithBool = new Pair<>(newBP, needAddnewExprWithState);
+                newExprWithState.add(newBPWithBool);
+                if (clauseType == ExprRewriter.ClauseType.ONCLAUSE) {
+                    analyzer.registerOnSlotToLiteralDeDuplication(pair);
+                    analyzer.registerOnSlotToLiteralExpr(newBP);
+                }
+            }
+        }
+    }
+
+    private void inferIsNotNullPredicates(List<Expr> slotEqSlotExpr,
+                                          List<Expr> isNullExpr,
+                                          Set<Expr> isNullDeDuplication,
+                                          List<Pair<Expr, Boolean>> newExprWithState,
+                                          Analyzer analyzer,
+                                          ExprRewriter.ClauseType clauseType) {
+        for (Expr isNullPredicate : isNullExpr) {
+            buildNewIsNotNullPredicate(isNullPredicate, slotEqSlotExpr, isNullDeDuplication,
+                    newExprWithState, analyzer, clauseType);
+        }
+    }
+
+    private void buildNewIsNotNullPredicate(Expr expr,
+                                            List<Expr> slotEqSlotExpr,
+                                            Set<Expr> isNullDeDuplication,
+                                            List<Pair<Expr, Boolean>> newExprWithState,
+                                            Analyzer analyzer,
+                                            ExprRewriter.ClauseType clauseType) {
+        if (expr instanceof IsNullPredicate) {
```

Comment:
We should check `IsNullPredicate.isSlotRefChildren` here.

Reply:
already edited